rts/Lua/LuaMatrixImpl.h

```diff
@@ -0,0 +1,164 @@
+	bool UnitMatrix(const unsigned int unitID, const sol::optional<bool> mult, sol::this_state L);
+	bool UnitPieceMatrix(const unsigned int unitID, const unsigned int pieceNum, const sol::optional<bool> mult, sol::this_state L);
+
+	bool FeatureMatrix(const unsigned int featureID, const sol::optional<bool> mult, sol::this_state L);
+	bool FeaturePieceMatrix(const unsigned int featureID, const unsigned int pieceNum, const sol::optional<bool> mult, sol::this_state L);
+
+	bool ProjectileMatrix(const unsigned int projID, const sol::optional<bool> mult, sol::this_state L);
+
+	void ScreenViewMatrix(const sol::optional<bool> mult);
+	void ScreenProjMatrix(const sol::optional<bool> mult);
+	void ScreenViewProjMatrix(const sol::optional<bool> mult);
+
+	void SunViewMatrix(const sol::optional<bool> mult);
+	void SunProjMatrix(const sol::optional<bool> mult);
+	void SunViewProjMatrix(const sol::optional<bool> mult);
+
+	void Ortho(const float left, const float right, const float bottom, const float top, const float near, const float far, const sol::optional<bool> mult);
+	void Frustum(const float left, const float right, const float bottom, const float top, const float near, const float far, const sol::optional<bool> mult);
+
+	void CameraViewMatrix(const sol::optional<unsigned int> cameraIdOpt, const sol::optional<bool> mult);
+	void CameraProjMatrix(const sol::optional<unsigned int> cameraIdOpt, const sol::optional<bool> mult);
+	void CameraViewProjMatrix(const sol::optional<unsigned int> cameraIdOpt, const sol::optional<bool> mult);
+	void CameraViewInverseMatrix(const sol::optional<unsigned int> cameraIdOpt, const sol::optional<bool> mult);
+	void CameraProjInverseMatrix(const sol::optional<unsigned int> cameraIdOpt, const sol::optional<bool> mult);
+	void CameraViewProjInverseMatrix(const sol::optional<unsigned int> cameraIdOpt, const sol::optional<bool> mult);
+	void CameraBillboardMatrix(const sol::optional<unsigned int> cameraIdOpt, const sol::optional<bool> mult);
+
+	tuple16f GetAsScalar()
+	{
+		return std::make_tuple(
+			mat[0], mat[1], mat[2], mat[3],
+			mat[4], mat[5], mat[6], mat[7],
+			mat[8], mat[9], mat[10], mat[11],
+			mat[12], mat[13], mat[14], mat[15]
+		);
+	}
+
+	sol::as_table_t<std::array<float, 16>> GetAsTable()
+	{
+		return sol::as_table(std::array<float, 16> {
+			mat[0], mat[1], mat[2], mat[3],
+			mat[4], mat[5], mat[6], mat[7],
+			mat[8], mat[9], mat[10], mat[11],
+			mat[12], mat[13], mat[14], mat[15]
+		});
+	}
+
+public:
+	const CMatrix44f& GetMatRef() const { return  mat; }
+	const CMatrix44f* GetMatPtr() const { return &mat; }
+public:
+	sol::as_table_t<std::array<float, 4>> operator* (const sol::table& tbl) const {
+		float4 f4 {0.0f, 0.0f, 0.0f, 1.0f};
+		for (auto i = 1; i <= 4; ++i) {
+			const sol::optional<float> maybeValue = tbl[i];
+			if (!maybeValue)
+				continue;
+
+			f4[i - 1] = maybeValue.value();
+		}
+		f4 = mat * f4;
+		return sol::as_table(std::array<float, 4> {f4.x, f4.y, f4.z, f4.w});
```

Comment:
why can't we use our types: float4, float3, CMatrix44f instead of std::array? seems like a lot of unnecessary copy-pasting going around

Reply:
I implemented the requested change, but unfortunately in this case values are stored as usertypes.
I'm gonna try to get some advise how to deal with it, but for now I'm reverting this change

Reply:
Are you sure about that? I thought this would trigger the container&& overload and basically copy the content to a table. My guess is we'd need to satisfy some sort of API (e.g. std::size and overload operator[] or provide a std::begin and std::end implementations)

Reply:
> Are you sure about that?

Checked this from Lua, The use of `float4` and `CMatrix44f` caused the creation of the userdata object.

> My guess is we'd need to satisfy some sort of API (e.g. std::size and overload operator[] or provide a std::begin and std::end implementations)

Looks like it's indeed the case. I'm sure at least about `.size()` requirement as I tried to also expose CMatrix44f.m and got `.size()` missing as one of the compilation errors.
Any idea what to do?

Reply:
changing `sol::as_table_t<std::array<float, 4>>` to `sol::as_table_t<float4>` in a few places (this function & in LuaMatris.cpp's sol:meta_function part (line 80)) is enough to make it compile: https://godbolt.org/z/9ahEqf . spring crashes on start due to that other issue you mentioned
The same can be done for the GetAsTable, like so:
```
	sol::as_table_t<CMatrix44f> GetAsTable()
	{
		return sol::as_table(mat);
	}
```

but I'm surprised this works as CMatrix44f doesn't seem to satisfy their requirement for a container (https://sol2.readthedocs.io/en/latest/api/as_table.html): "The container passed here can be a pointer, a reference, a std::reference_wrapper around a container, or just a plain container value. It must have a begin/end function, and if it has a std::pair<Key, Value> as its value_type, it will be pushed as a dictionary. Otherwise, it’s pushed as a sequence."

It doesn't seem to have an .begin() and .end() function. If it turns out that things are just compiling, but not working, I recommend you add a .begin() and .end() functions that would point to the first and after-last member of the matrix and vector